15191-Undeclared-cleanOutUndeclared-should-not-iterate-over-all-methods

MarcusDenker opened the pull request:
This PR now does:

- Inline and remove the private method #declare:from: from Dictionaries
- This allows to make sure that when we move variables to and from Undeclared, we correctly change the class (this code is not nice, we need to improve the Undeclared fixing in a second step)
- fix tests to not put a Association into the Undeclared Dict
- fix SystemDictionary>>#at:put: to no use Associations when using #at:put: 

And the original starting point:
- do not iterate all methods for #cleanOutUndeclared, but use the recorded #usingMethods


fixes #15191
fixes #11903

MarcusDenker committed:
Undeclared variables now know the methods they are used in.

We can this clean up Undeclared without having to iterate over all methods

fixes #15191

MarcusDenker commented:
Failing tests 

```
testClassObsolete – Unix64.ReleaseTests.CleanCode.ObsoleteTest
testFixObsoleteSharedPools – Unix64.ReleaseTests.CleanCode.ObsoleteTest
testTraitObsolete – Unix64.ReleaseTests.CleanCode.ObsoleteTest
testObsoleteClasses – Unix64.ReleaseTests.Release.ReleaseTest
testUndeclared – Unix64.ReleaseTests.Release.ReleaseTest
```

jecisc commented:
So we still have Associations in the SystemDictionary? This is weird!

jecisc commented:
Those tests are green in a clean image. This means that we have tests adding associations to the SystemDictionary! It's not the build of the image that is doing this.

MarcusDenker commented:
The associations might be in the Undeclared Dictionary

MarcusDenker committed:
- use #unregister in #cleanOutUndeclared
- fix a test that was adding an Associaton to the Undeclared Dictionay

jecisc commented:
```
Message not understood: GlobalVariable >> #unregister
Stacktrace
MessageNotUnderstood
Message not understood: GlobalVariable >> #unregister
GlobalVariable(Object)>>doesNotUnderstand: #unregister
[ :var | var unregister ] in SmalltalkImage>>cleanOutUndeclared
Array(SequenceableCollection)>>select:thenDo:
SmalltalkImage>>cleanOutUndeclared
SmalltalkImage>>fixObsoleteSharedPools
SmalltalkImage>>fixObsoleteReferences
ObsoleteTest>>tearDown
```

MarcusDenker committed:
- #removeFromSystem: should just register a new Undelared, not use #declare:from:
- simplfy testUndeclaredVariableWhenItIsAlreadyRegisteredAsAssociation

MarcusDenker committed:
-  inline #declare:from:, remove tests. Fixes #11903
- make sure that #removeFromSystem: and #removeClassVarNamed:interactive: change the variables to Undeclared
- make sure #at:put in SystemDictionary never uses Associations

MarcusDenker committed:
Merge branch 'Pharo12' into 15191-Undeclared-cleanOutUndeclared-should-not-iterate-over-all-methods

MarcusDenker committed:
Update Class.class.st

add guard in #removeFromSystem:
